crates/bevy_ecs/macros/src/fetch.rs

```diff
@@ -112,37 +112,59 @@ pub fn derive_world_query_impl(ast: DeriveInput) -> TokenStream {
 
     let state_struct_name = Ident::new(&format!("{struct_name}State"), Span::call_site());
 
-    let fields = match &ast.data {
-        Data::Struct(DataStruct {
-            fields: Fields::Named(fields),
-            ..
-        }) => &fields.named,
-        _ => panic!("Expected a struct with named fields"),
+    let Data::Struct(DataStruct { fields, .. }) = &ast.data else {
+        return syn::Error::new(
+            Span::call_site(),
+            "#[derive(WorldQuery)]` only supports structs",
+        )
+        .into_compile_error()
+        .into()
     };
+    if fields.is_empty() {
+        return syn::Error::new(
+            Span::call_site(),
+            "#[derive(WorldQuery)]` does not support fieldless structs",
```

Comment:
Currently, fieldless structs fail to compile due to an inscrutable macro lifetime error. This just makes that error more descriptive.

Ideally we would just fix that lifetime error, but I haven't found an elegant way to do that yet. Plus, it's not all that important to support fieldless structs (would be nice though).